File: Buoi2/vidu/bubble.cpp
#include <iostream>
#include <vector>
using namespace std;

// Hàm Bubble Sort
void bubble_sort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;  // Đánh dấu có hoán đổi hay không
        
        for (int j = 0; j < n - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                // Hoán đổi các phần tử
                swap(arr[j], arr[j + 1]);
                swapped = true;  // Đã có hoán đổi
            }
        }
        
        // Nếu không có hoán đổi nào, mảng đã được sắp xếp
        if (!swapped) {
            break;
        }
    }
}

int main() {
    vector<int> arr = {5, 2, 9, 1, 5, 6};  // Khởi tạo mảng
    bubble_sort(arr);  // Gọi hàm sắp xếp
    
    cout << "Kết quả sau Bubble Sort: ";
    for (int x : arr) {
        cout << x << " ";  // In kết quả
    }
    cout << endl;
    
    return 0;
}

File: Buoi2/vidu/bubble.jl
# Hàm Bubble Sort
function bubble_sort(arr)
    n = length(arr)
    for i in 1:n-1
        swapped = false  # Đánh dấu có hoán đổi hay không
        
        for j in 1:n-1-i
            if arr[j] > arr[j + 1]
                # Hoán đổi các phần tử
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = true  # Đã có hoán đổi
            end
        end
        
        # Nếu không có hoán đổi nào, mảng đã được sắp xếp
        if !swapped
            break
        end
    end
end

# Mảng đầu vào
arr = [5, 2, 9, 1, 5, 6]  # Khởi tạo mảng
bubble_sort(arr)  # Gọi hàm sắp xếp

# In kết quả
println("Kết quả sau Bubble Sort: ", arr)
